pkg/controller/direct/directbase/directbase_controller.go

```diff
@@ -0,0 +1,394 @@
+	logger := log.FromContext(ctx)
+
+	resource, err := toK8sResource(policy)
+	if err != nil {
+		logger.Error(err, "error converting to k8s resource while handling event",
+			"resource", k8s.GetNamespacedName(policy), "event", k8s.DeleteFailed)
+		return fmt.Errorf(k8s.DeleteFailedMessageTmpl, origErr)
+	}
+	return r.Reconciler.HandleDeleteFailed(ctx, resource, origErr)
+}
+
+func (r *DirectReconciler) supportsImmediateReconciliations() bool {
+	return r.immediateReconcileRequests != nil
+}
+
+func (r *reconcileContext) handleUnresolvableDeps(ctx context.Context, policy *unstructured.Unstructured, origErr error) (requeue bool, err error) {
+	logger := log.FromContext(ctx)
+
+	resource, err := toK8sResource(policy)
+	if err != nil {
+		return false, fmt.Errorf("error converting to k8s resource while handling unresolvable dependencies event: %w", err)
+	}
+	refGVK, refNN, ok := lifecyclehandler.CausedByUnreadyOrNonexistentResourceRefs(origErr)
+	if !ok || !r.Reconciler.supportsImmediateReconciliations() {
+		// Requeue resource for reconciliation with exponential backoff applied
+		return true, r.Reconciler.HandleUnresolvableDeps(ctx, resource, origErr)
+	}
+	// Check that the number of active resource watches
+	// does not exceed the controller's cap. If the
+	// capacity is not exceeded, The number of active
+	// resource watches is incremented by one and a watch
+	// is started
+	if !r.Reconciler.resourceWatcherRoutines.TryAcquire(1) {
+		// Requeue resource for reconciliation with exponential backoff applied
+		return true, r.Reconciler.HandleUnresolvableDeps(ctx, resource, origErr)
+	}
+	// Create a logger for ResourceWatcher that contains info
+	// about the referencing resource. This is done since the
+	// messages logged by ResourceWatcher only include the
+	// information of the resource it is watching by default.
+	watcherLogger := logger.WithValues(
+		"referencingResource", resource.GetNamespacedName(),
+		"referencingResourceGVK", resource.GroupVersionKind())
+	watcher, err := resourcewatcher.New(r.Reconciler.config, watcherLogger)
+	if err != nil {
+		return false, r.Reconciler.HandleUpdateFailed(ctx, resource, fmt.Errorf("error initializing new resourcewatcher: %w", err))
+	}
+
+	logger = logger.WithValues(
+		"resource", resource.GetNamespacedName(),
+		"resourceGVK", resource.GroupVersionKind(),
+		"reference", refNN,
+		"referenceGVK", refGVK)
+	go func() {
+		// Decrement the count of active resource watches after
+		// the watch finishes
+		defer r.Reconciler.resourceWatcherRoutines.Release(1)
+		timeoutPeriod := jitter.GenerateWatchJitteredTimeoutPeriod()
+		ctx, cancel := context.WithTimeout(context.TODO(), timeoutPeriod)
+		defer cancel()
+		logger.Info("starting wait with timeout on resource's reference", "timeout", timeoutPeriod)
+		if err := watcher.WaitForResourceToBeReady(ctx, refNN, refGVK); err != nil {
```

Comment:
Doing this under a weighted semaphore seems like it may cause (?exacerbate) ordering problems? Holding a semaphore, even a weighted semaphore and then blocking indefinitely just seems to be asking for problems. Are we sure this isn't a premature optimization? I've seen KAS and KCM  ticking overly fairly happily with hundreds of thousands of goroutines. I'd like to be convinced we actually need to solve this issue. If we do, then I wonder if this is the best way to do it.

Reply:
So good news / bad news ... directbase_controller started with a copy of one of our existing go controllers (iampolicy_controller.go, IIRC).  I don't think I ever looked at this code, it wasn't in the critical path for the direct controllers yet :-)  I agree with you that we can do better, we actually have the logic to do so in our current experimental/composite directory (TrackingDynamicClient). 

Do you mind filing a github issue so we don't lose this problem, but I don't think we want to solve it in this PR.